src/base/rss/rss_autodownloader.cpp

```diff
@@ -226,23 +226,25 @@ void AutoDownloader::importRules(const QByteArray &data, AutoDownloader::RulesFi
 QByteArray AutoDownloader::exportRulesToJSONFormat() const
 {
     QJsonObject jsonObj;
-    for (const auto &rule : rules())
+    const auto rulesList = rules();
+    for (const auto &rule : rulesList)
```

Comment:
This form is bad IMO.

Is it possible to introduce a helper for qt containers?
```c++
template <typename T>
constexpr typename std::add_const<T>::type qConstify(const T &t)
{
    return t;
}

// usage:
for (const auto &rule : qConstify(rulesList)) {}
```
To my knowledge, the `constify()` will then return a temporary const object and so no detach, right?

Reply:
I thought about it too. I would prefer to mix qAsConst (`std::as_const`) behavior with @Chocobo1 suggested one so it returns const copy for `rvalue`'s and casts to const for `lvalue`'s.

Reply:
@Chocobo1 You meant:
```
for (const auto &rule : qConstify(rules())) {}
```
right ?

Reply:
>for (const auto &rule : qConstify(rules())) {}
>
>right ?

Exactly, was a copy-paste error.

Reply:
I think he means it exactly.

Reply:
Ok.
This is a semantic question, IMO. My rationale can be more or less summarized in this post: https://www.mail-archive.com/development@qt-project.org/msg30433.html

Reply:
>This is a semantic question, IMO. My rationale can be more or less summarized in this post: https://www.mail-archive.com/development@qt-project.org/msg30433.html

That is viewpoint from library, but we are at application level, that is the main difference.

IMO, the performance price to pay is marginal and we gain huge in readability & maintainability (given @glassez proposition and mine).

Reply:
@Chocobo1 @glassez 
My proposal:
```cpp
template <typename T>
Q_DECL_CONSTEXPR typename std::add_const<T>::type &qBtConstify(T &t) Q_DECL_NOTHROW { return t; }

template <typename T>
Q_DECL_CONSTEXPR typename std::add_const<T>::type qBtConstify(T &&t) Q_DECL_NOTHROW { return std::move(t); }
```

I actually like `qConstify` but it might be taken as a native Qt thing. Renamed to `qBtConstify`.

Reply:
1. c++11 is mandatory for qbt, you can use `constexpr` & `noexcept` directly.
2. personally I would prefer `qBtConstify(const T &t)`, albeit there is not much meaning of it. 
3. the helper might can be placed in "src\base\utils", not sure which file yet.

Reply:
1. Didn't check what where qbt requirements. Will use the standard keywords.
2. I also thought about. But (a) as you said there's not much meaning in it. (b) This way it follows `std::as_const()`.
3. I was thinking in `src/base/global.h`